packages/notebook/src/executionindicator.tsx

```diff
@@ -0,0 +1,528 @@
+     */
+    attachNotebook(
+      data: { content?: Notebook; context?: ISessionContext } | null
+    ): void {
+      if (data && data.content && data.context) {
+        const nb = data.content;
+        const context = data.context;
+        this._currentNotebook = nb;
+        if (!this._notebookExecutionProgress.has(nb)) {
+          this._notebookExecutionProgress.set(nb, {
+            executionStatus: 'idle',
+            kernelStatus: 'idle',
+            totalTime: 0,
+            interval: 0,
+            timeout: 0,
+            scheduledCell: new Set<string>(),
+            scheduledCellNumber: 0,
+            needReset: true
+          });
+
+          const state = this._notebookExecutionProgress.get(nb);
+          context.statusChanged.connect(ctx => {
+            if (state) {
+              state.kernelStatus = ctx.kernelDisplayStatus;
+            }
+            this.stateChanged.emit(void 0);
+          }, this);
+
+          context.connectionStatusChanged.connect(ctx => {
+            if (state) {
+              state.kernelStatus = ctx.kernelDisplayStatus;
+            }
+            this.stateChanged.emit(void 0);
+          }, this);
+
+          const handleKernelMsg = (
+            sender: IKernelConnection,
+            msg: IAnyMessageArgs
+          ) => {
+            const message = msg.msg;
+            const msgId = message.header.msg_id;
+
+            if (
+              KernelMessage.isCommMsgMsg(message) &&
+              message.content.data['method']
+            ) {
+              // Execution request from Comm message
+              const method = message.content.data['method'];
+              if (method !== 'request_state' && method !== 'update') {
+                this._cellScheduledCallback(nb, msgId);
+                this._startTimer(nb);
+              }
+            } else if (message.header.msg_type === 'execute_request') {
+              // A cell code is scheduled for executing
+              this._cellScheduledCallback(nb, msgId);
+            } else if (
+              KernelMessage.isStatusMsg(message) &&
+              message.content.execution_state === 'idle'
+            ) {
+              // Idle status message case.
+              const parentId = (message.parent_header as KernelMessage.IHeader)
+                .msg_id;
+              this._cellExecutedCallback(nb, parentId);
+            } else if (message.header.msg_type === 'execute_input') {
+              // A cell code starts executing.
+              this._startTimer(nb);
+            }
+          };
+          context.session?.kernel?.anyMessage.connect(handleKernelMsg);
+          context.kernelChanged.connect((_, kernelData) => {
+            if (state) {
+              this._resetTime(state);
+              this.stateChanged.emit(void 0);
+              if (kernelData.newValue) {
+                kernelData.newValue.anyMessage.connect(handleKernelMsg);
+              }
+            }
+          });
```

Comment:
All slots connected to a kernel aren't deleted when this kernel is disposed?

Reply:
I just take a look at `Popup` widget, since in my case,  the popup is displayed  on hover instead of on click  and  its position is on top or bottom of the indicator, depend on the position of indicator (toolbar or status bar),  it would need a good amount of  modifications to incorporate these changes.

Reply:
> All slots connected to a kernel aren't deleted when this kernel is disposed?

It is worth checking ;)

>  it would need a good amount of modifications to incorporate these changes.

Ok then let's not use `Popup` then.

Reply:
Thanks